irobotclient/request_handler.py

```diff
@@ -0,0 +1,161 @@
+is not going to work.
+"""
+REQUEST_LIMIT = 10
+
+"""
+If a 202 response returns with not iRobot-ETA header then a default delay (in seconds) will be set.
+"""
+DEFAULT_REQUEST_DELAY = 600
+
+"""
+The iRobot response status codes are declared below along with an associated standard error number if applicable.
+"""
+ResponseStruct = namedtuple("ResponseStruct", "status_code, errno")
+
+RESPONSES = {
+    'SUCCESS': ResponseStruct(status_code=200, errno=None),
+    'FETCHING_DATA': ResponseStruct(status_code=202, errno=None),
+    'RANGED_DATA': ResponseStruct(status_code=206, errno=None),
+    'CLIENT_MATCHED': ResponseStruct(status_code=304, errno=None),
+    'AUTH_FAIL': ResponseStruct(status_code=401, errno=errno.ECONNREFUSED),
+    'DENIED_IRODS': ResponseStruct(status_code=403, errno=errno.EACCES),
+    'NOT_FOUND': ResponseStruct(status_code=404, errno=errno.ENODATA),
+    'INVALID_REQUEST_METHOD': ResponseStruct(status_code=405, errno=errno.EPROTO),
+    'INVALID_MEDIA_REQUESTED': ResponseStruct(status_code=406, errno=errno.EINVAL),
+    'INVALID_RANGE': ResponseStruct(status_code=416, errno=errno.ERANGE),
+    'TIMEOUT': ResponseStruct(status_code=504, errno=errno.ETIMEDOUT),
+    'PRECACHE_FULL': ResponseStruct(status_code=507, errno=errno.ENOMEM)
+}
+
+
+class Requester:
+    def __init__(self, requested_url: str, headers: dict):
+        """
+
+        :param requested_url:
+        :param headers:
+        """
+
+        self._request = requests.Request(url=requested_url, headers=headers)
+        self._request_delay = 0
+
+    def get_data(self) -> requests.Response:
+        """
+
+        :return:
+        """
+        try:
+            for index in range(REQUEST_LIMIT):
+
+                time.sleep(self._request_delay)
+
+                response = requests.head(self._request)
+
+                if response.status_code == RESPONSES['SUCCESS'].status_code:
+                    return requests.get(self._request_delay, stream=True)
+
+                elif response.status_code == RESPONSES['FETCHING_DATA'].status_code:
+                    self._request_delay = self._get_request_delay(response)
+
+                elif response.status_code == RESPONSES['RANGED_DATA'].status_code:
+                    pass  # TODO - This response could have a ETA of remaining data ranges
+
+                elif response.status_code == RESPONSES['CLIENT_MATCHED'].status_code:
+                    pass  # TODO - Client has already downloaded this data; need to add sum to request for this to work?
+
+                elif response.status_code == RESPONSES['AUTH_FAIL'].status_code:
+                    pass  # TODO - Check for response expected authorisation type; try basic auth.
+
+                elif response.status_code == RESPONSES['DENIED_IRODS'].status_code:
+                    raise IrobotClientException(errno=RESPONSES['DENIED_IRODS'].errno,
+                                                message="ERROR: Access to IRODs denied.")
```

Comment:
It's more like "Access denied by iRODS": they user can access iRODS, but the file they're trying to get hold of is restricted in some way.